sql: introduce special type for inverted index keys

RaduBerinde opened the pull request:
Note: this is intended to merge to master after the 22.1 branch is cut.

Currently we pass around inverted index keys as having `Bytes` type,
but these datums behave in subtly different ways. The
encoding/decoding is special-cased to be a no-op. This leads to sharp
edges; for example, we put these keys in `EncodedDatum`s but if we
ever try to decode them (via `EnsureDecodedd`), we get obscure
decoding errors.

This commit cleans up this situation by adding a special `EncodedKey`
type and corresponding `DEncodedKey` datum. It behaves like
`Bytes/DBytes` in most cases, except key encoding and decoding.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/77501)
<!-- Reviewable:end -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 27 of 27 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77501#-:-MxfPRSB0JqWBMX1zHlN:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @RaduBerinde, and @rytaft)

___
*[-- commits, line 8 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77501#-MxfNpUNAeLj5zS6TPGv):*
nit: `s/Decodedd/Decoded/`.
___
*[pkg/sql/sem/tree/datum.go, line 1626 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77501#-MxfOpXA0hxksaj-C4I6:-MxfOpXA0hxksaj-C4I7:b6cpppx) ([raw file](https://github.com/cockroachdb/cockroach/blob/faf90d109bb9c3616791da1bbedcf84d93ee43d8/pkg/sql/sem/tree/datum.go#L1626)):*
> ```Go
> // Format implements the NodeFormatter interface.
> func (d *DEncodedKey) Format(ctx *FmtCtx) {
> 	f := ctx.flags
> ```

nit: it seems like it'd be good to extract a helper to be shared by two Bytes datums.
___
*[pkg/sql/sem/tree/datum_alloc.go, line 151 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77501#-MxfP0yKBK7OkWTXu1Kn:-MxfP0yKBK7OkWTXu1Ko:bmfqm0e) ([raw file](https://github.com/cockroachdb/cockroach/blob/faf90d109bb9c3616791da1bbedcf84d93ee43d8/pkg/sql/sem/tree/datum_alloc.go#L151)):*
> ```Go
> }
> 
> // NewDEncodedKey allocates a DBytes.
> ```

nit: `s/DBytes/DEncodedKey/`.


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
Great idea!

To prevent poor row count estimations I think we'll need to include the new type you've created here: https://github.com/cockroachdb/cockroach/blob/9b385e625a30eff950bbb10c561809744eeb6a8c/pkg/sql/opt/props/histogram.go#L835

Otherwise, if a query filters by a value in the middle of a histogram bucket, the estimated row count will be the bucket's `num_range` divided by two.

Does the `histo_col_type` field need to change in table statistics? Or is it fine to leave that as `BYTES` for inverted index columns.

> Reviewed 27 of 27 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77501#-:-MxfVoH48mIIrOe6knjQ:bxz0knl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/types/types.go, line 468 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77501#-MxfTmUo7qxZ35erh2OC:-MxfTmUo7qxZ35erh2OD:b3peooc) ([raw file](https://github.com/cockroachdb/cockroach/blob/faf90d109bb9c3616791da1bbedcf84d93ee43d8/pkg/sql/types/types.go#L468)):*
> ```Go
> 	// EncodedKey is a special type used internally for inverted index keys, which
> 	// do not fully encode an object.
> 	EncodedKey = &T{
> ```

nit: Unless you think it's likely this will be used for other purposes in the future, it might make sense to make the name of the type and family more specific to inverted index keys.


<!-- Sent from Reviewable.io -->

mgartner commented:
> To prevent poor row count estimations...

See https://github.com/cockroachdb/cockroach/pull/68740 for some more context on this.





<!-- Sent from Reviewable.io -->

mgartner commented:
> To prevent poor row count estimations...

Hmm, I guess these datums are the upper and lower bound of the histogram buckets, so if the type remains as `BYTES`, perhaps no change is needed here.

RaduBerinde reviewed (commented):
The statistics creation and props code still uses DBytes (`props.makeSpanFromInvertedSpan`), I don't think it's worth changing that. I will add some more comments around the code though.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77501#-:-Mxjd1QX5brweoy0_xdi:bt5abyq)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @RaduBerinde, and @rytaft)

___
*[pkg/sql/types/types.go, line 468 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77501#-MxfTmUo7qxZ35erh2OC:-MxjdHNCAaF8eDTWf1Dw:bkmwmjh) ([raw file](https://github.com/cockroachdb/cockroach/blob/faf90d109bb9c3616791da1bbedcf84d93ee43d8/pkg/sql/types/types.go#L468)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: Unless you think it's likely this will be used for other purposes in the future, it might make sense to make the name of the type and family more specific to inverted index keys.
</blockquote></details>

I think I'd like to keep it, it's a sound concept of its own, maybe there will be some use case for passing around encoded keys. I could imagine potential usecases around composite types like collated strings.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77501)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @rytaft, and @yuzefovich)

___
*[pkg/sql/sem/tree/datum.go, line 1626 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77501#-MxfOpXA0hxksaj-C4I6:-MxlxwOX7eyEb7ya2HAb:b-bxum73) ([raw file](https://github.com/cockroachdb/cockroach/blob/faf90d109bb9c3616791da1bbedcf84d93ee43d8/pkg/sql/sem/tree/datum.go#L1626)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: it seems like it'd be good to extract a helper to be shared by two Bytes datums.
</blockquote></details>

I just "rerouted" to the DBytes implementation.


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> The statistics creation and props code still uses DBytes (props.makeSpanFromInvertedSpan), I don't think it's worth changing that. I will add some more comments around the code though.

SGTM. Thanks!

> Reviewed 4 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77501#-:-MxoLpuYDxiDn7OLHCoa:bu5c8x3)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @RaduBerinde, @rytaft, and @yuzefovich)

___
*[pkg/sql/sem/tree/datum.go, line 1564 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/77501#-MxoL9B05fK2HVF_n0vu:-MxoL9B05fK2HVF_n0vv:bu3sy0k) ([raw file](https://github.com/cockroachdb/cockroach/blob/49928f3cc0d41309a9d8ba8d43d68102a6b03ed8/pkg/sql/sem/tree/datum.go#L1564)):*
> ```Go
> }
> 
> // DEncodedKey is the bytes Datum. The underlying type is a string because we want
> ```

nit: `DEncodedKey` is _not_ the bytes datum. It'd be helpful to explain the contrast between the type and `DBytes` here.
___
*[pkg/sql/types/types.go, line 468 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77501#-MxfTmUo7qxZ35erh2OC:-MxoK0sh13yu3-z6hS5G:b-8jghhs) ([raw file](https://github.com/cockroachdb/cockroach/blob/faf90d109bb9c3616791da1bbedcf84d93ee43d8/pkg/sql/types/types.go#L468)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I think I'd like to keep it, it's a sound concept of its own, maybe there will be some use case for passing around encoded keys. I could imagine potential usecases around composite types like collated strings.
</blockquote></details>

Sounds good to me. In that case, I think the comment should be more general and explain the behavior of the type (e.g. "EncodedKey is similar to DBytes but has a different behavior for encoding and decoding ...") before mentioning the inverted index key use-case.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
TFTRs!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77501#-:-Mxp--EbD59Z1gGJtUKv:bz5mair)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @RaduBerinde, @rytaft, and @yuzefovich)

___
*[pkg/sql/types/types.go, line 468 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77501#-MxfTmUo7qxZ35erh2OC:-MxozvG43Lo_4gfzkYMi:b-oq3912) ([raw file](https://github.com/cockroachdb/cockroach/blob/faf90d109bb9c3616791da1bbedcf84d93ee43d8/pkg/sql/types/types.go#L468)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Sounds good to me. In that case, I think the comment should be more general and explain the behavior of the type (e.g. "EncodedKey is similar to DBytes but has a different behavior for encoding and decoding ...") before mentioning the inverted index key use-case.
</blockquote></details>

Done, PTAL


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 5 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77501#-:-Mxq2R6N1dh87pg3mTWy:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @rytaft and @yuzefovich)

___
*[pkg/sql/types/types.go, line 468 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77501#-MxfTmUo7qxZ35erh2OC:-Mxq2L_G9T20i8Jx86zK:byx7juc) ([raw file](https://github.com/cockroachdb/cockroach/blob/faf90d109bb9c3616791da1bbedcf84d93ee43d8/pkg/sql/types/types.go#L468)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Done, PTAL
</blockquote></details>

Looks great thanks!


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
sql: introduce special type for inverted index keys

Currently we pass around inverted index keys as having `Bytes` type,
but these datums behave in subtly different ways. The
encoding/decoding is special-cased to be a no-op. This leads to sharp
edges; for example, we put these keys in `EncodedDatum`s but if we
ever try to decode them (via `EnsureDecoded`), we get obscure
decoding errors.

This commit cleans up this situation by adding a special `EncodedKey`
type and corresponding `DEncodedKey` datum. It behaves like
`Bytes/DBytes` in most cases, except key encoding and decoding.

Release note: None

RaduBerinde commented:
bors r+
